backend/api_calls.py

```diff
@@ -877,6 +877,132 @@ def RandomWords(n):
+    raise InvalidInputError('sendTime must not be in the past!')
+
+  if 'previewMessage' not in request:
+    request['previewMessage'] = textwrap.wrap(request['message'], 100)[0]
+
+  put_data = {}
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon', 'previewMessage']
+
+  for property in properties:
+    if property in request:
+      put_data[property] = request[property]
+
+  return firebase.put('/notifications',
+                      request['notificationId'], put_data)
+
+
+def UpdateNotification(request, firebase):
+  """Update a queued notification.
+
+  Validation:
+    notificationId exists.
+
+  Args:
+    notificationId: Queued Notification id to create.
+    message: Message to send to client.
+    previewMessage: Short preview of the message conents. Optional.
+    app: Whether to send as a push notification to the iOS/Android/Web clients.
+    vibrate: Whether the notification should vibrate on iOS/Android.
+    sound: Whether the notification should play a sound on iOS/Android.
+    destination: URL that notification should open when clicked.
+    sendTime: When to spawn notifications from this template. UNIX timestamp.
+    playerId: Player id to send message to.
+    groupId: Who to send the notifications to.
+    icon: An icon code to show.
+
+  Firebase entries:
+    /notifications/%(notificationId)
+"""
+  valid_args = ['notificationId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+
+  put_data = {}
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon']
+
+  current_time = time.time()
+  if 'sendTime' in request and current_time > int(request['sendTime']):
+    raise InvalidInputError('sendTime must not be in the past!')
+
+  notification = firebase.get('/notifications', request['notificationId'])
+  # This shouldn't happen since we validated this above...
+  if not notification:
+    raise InvalidInputError('notificationId must exist!')
+  if current_time > int(notification['sendTime']) or 'sent' in notification:
+    raise InvalidInputError('Cannot modify sent notification.')
+
+  for property in properties:
+    if property in request:
+      put_data[property] = request[property]
+
+  return firebase.patch('/notifications/%s' % request['notificationId'], put_data)
+
+
+def MarkNotificationSeen(request, firebase):
+  """Updates the notification's seenTime.
+
+  Validation:
+    notificationId must exist.
+    playerId must exist.
+
+  Args:
+    notificationId: Notification Id to update.
+    playerId: Player who saw notification.
+
+  Firebase entries:
+    /players/%(playerId)/notifications/%(notificationId)
+"""
+  valid_args = ['notificationId', 'playerId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+  put_data = {
+    'time': time.time()
```

Comment:
Note: time.time() returns a float, not a UNIX timestamp. Not sure if we should stick to timestamps or not.
For timestamp, int(time.time())